test/blocklib/core/fft/fft.hpp

```diff
@@ -1,153 +1,63 @@
-    template<typename evalU> struct EvalOutputType { using type1 = evalU; using type2 = typename MakeSigned<T>::type;};
-    template<ComplexType evalU> struct EvalOutputType<evalU> { using type1 = typename evalU::value_type; using type2 = evalU;};
-    using U = std::conditional_t<ComplexType<T>, typename EvalOutputType<T>::type1, typename EvalOutputType<T>::type2>;
-    // clang-format on
-
-    using InDataType    = typename fftw<T>::InDataType;
-    using OutDataType   = typename fftw<T>::OutDataType;
-    using PrecisionType = typename fftw<T>::PrecisionType;
-    using InHistoryType = typename fftw<T>::InHistoryType;
-    using InUniquePtr   = typename fftw<T>::InUniquePtr;
-    using OutUniquePtr  = typename fftw<T>::OutUniquePtr;
-    using PlanUniquePtr = typename fftw<T>::PlanUniquePtr;
-
-    PortIn<T>                     in;
-    PortOut<DataSet<U>>           out;
-
-    std::size_t                   fftSize{ 1024 };
-    int                           window{ static_cast<int>(WindowFunction::None) };
-    std::vector<PrecisionType>    windowVector;
-    bool                          outputInDb{ false };
-    std::string                   wisdomPath{ ".gr_fftw_wisdom" };
-    int                           sign{ FFTW_FORWARD };
-    unsigned int                  flags{ FFTW_ESTIMATE }; // FFTW_EXHAUSTIVE, FFTW_MEASURE, FFTW_ESTIMATE
-    history_buffer<InHistoryType> inputHistory{ fftSize };
-    InUniquePtr                   fftwIn{};
-    OutUniquePtr                  fftwOut{};
-    PlanUniquePtr                 fftwPlan{};
-    std::vector<U>                magnitudeSpectrum{};
-    std::vector<U>                phaseSpectrum{};
-
-    fft() { initAll(); }
-
-    fft(const fft &rhs)     = delete;
-    fft(fft &&rhs) noexcept = delete;
-    fft &
-    operator=(const fft &rhs)
+    using U             = gr::algorithm::FFTOutputType<T>;
+    using InDataType    = typename FFTInternal<T>::InDataType;
+    using OutDataType   = typename FFTInternal<T>::OutDataType;
+    using PrecisionType = typename FFTInternal<T>::PrecisionType;
+
+    PortIn<T>                                                               in;
+    PortOut<DataSet<U>>                                                     out;
+
+    TImpl                                                                   fftImpl;
+    Annotated<std::size_t, "FFT size", Doc<"FFT size">>                     fftSize{ 1024 };
+    Annotated<int, "window function", Doc<"window (apodization) function">> window{ static_cast<int>(WindowFunction::Hann) };
+    Annotated<float, "sample rate", Doc<"signal sample rate">, Unit<"Hz">>  sampleRate = 1.f;
+    Annotated<bool, "output in dB", Doc<"calculate output in decibels">>    outputInDb{ false };
+    std::vector<PrecisionType>                                              windowVector;
+    history_buffer<InDataType>                                              inputHistory{ fftSize };
```

Comment:
not needed/inefficient -> propose to drop in favour of using `process_bulk(...)`

N.B. you need to set the min/max `RequiredSamples<>`of the input port to the required default fftSize (i.e. 1024) during the init process and change if the user changes the settings.